cpp/src/shap/kernel_shap.cu

```diff
@@ -0,0 +1,281 @@
+          }
+        }
+      }
+  }
+
+
+
+
+template <typename DataT, typename IdxT>
+__global__ void sampled_rows_kernel(IdxT nsamples,
+                                    int* X,
+                                    IdxT nrows_X,
+                                    IdxT M,
+                                    DataT* background,
+                                    IdxT nrows_background,
+                                    DataT* combinations,
+                                    DataT* observation){
+  extern __shared__ int smps[];
+  int tid = threadIdx.x + blockDim.x * blockIdx.x;
+  int i, j;
+
+  if(threadIdx.x < nrows_background){
+    // thread 0 of block generates samples, reducing number of rng calls
+    // calling curand only 3 * nsamples times.
+    // Sampling algo from: Li, Kim-Hung. "Reservoir-sampling algorithms
+    // of time complexity O (n (1+ log (N/n)))." ACM Transactions on Mathematical
+    // Software (TOMS) 20.4 (1994): 481-493.
+    if(threadIdx.x == 0){
+      float w;
+      curandState_t state;
+      for(i=0; i<nsamples; i++){
+        smps[i] = i;
+      }
+      curand_init((unsigned long long)clock64(),
+                  (unsigned long long)tid,
+                  0,
+                  &state);
+
+      w = exp(log(curand_uniform(&state)) / nsamples);
+
+      while(i < M){
+        i = i + floor(log(curand_uniform(&state)) / log(1 - w)) + 1;
+        if(i <= M){
+          smps[(int)(curand_uniform(&state) * nsamples)] = i;
+          w = w * exp(log(curand_uniform(&state)) / nsamples);
+        }
+      }
+
+      // write samples to 1-0 matrix
+      for(i=0; i<nsamples; i++){
+        X[i] = smps[i];
+      }
+    }
+
+
+    // all threads write background line to their line
+#pragma unroll
+    for(i=tid; i<nrows_background; i+=blockDim.x){
+#pragma unroll
+      for(j=0; j<M; j++){
+        combinations[tid * M + j] = background[blockIdx.x * M + j];
+      }
+    }
+
+    __syncthreads();
+
+    // all threads write observation[samples] into their entry
+#pragma unroll
+    for(i=tid; i<nrows_background; i+=blockDim.x){
+#pragma unroll
+      for(j=0; j<nsamples; j++){
+        combinations[tid * M + smps[i]] = observation[smps[j]];
+      }
+    }
+  }
+}
+
+
+template <typename DataT, typename IdxT>
+void kernel_dataset_impl(const raft::handle_t& handle,
+                         int* X,
+                         IdxT nrows_X,
+                         IdxT M,
+                         DataT* background,
+                         IdxT nrows_background,
+                         DataT* combinations,
+                         DataT* observation,
+                         int* nsamples,
+                         int len_nsamples,
+                         int maxsample){
+    const auto& handle_impl = handle;
+    cudaStream_t stream = handle_impl.get_stream();
+
+    IdxT nblks;
+    IdxT Nthreads;
+    Nthreads = int(32 / nrows_background) * 32
+
+    if(M * sizeof(DataT) <= 49152){
```

Comment:
we should have a call somewhere to obtain shared memory size if this is the constant... maybe cudaFuncGetAttribute can do it? A100 should be able to support much larger shared memory.
 Might be nice if we issued a debug-level log to indicate which path we're taking since it will affect perf.